sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -203,223 +156,218 @@ public BoundedJdbcSource withPassword(String password) {
     @Nullable
     private final String password;
 
-    public BoundedJdbcSource(DataSource dataSource, String query, String username,
-                             String password) {
+    private Read(DataSource dataSource, String query, @Nullable String username,
+                 @Nullable String password) {
       this.dataSource = dataSource;
       this.query = query;
       this.username = username;
       this.password = password;
     }
 
     @Override
-    public Coder getDefaultOutputCoder() {
-      return SerializableCoder.of(JdbcDataRecord.class);
+    public PCollection<JdbcDataRecord> apply(PBegin input) {
+      JdbcOptions jdbcOptions = JdbcOptions.from(dataSource, query, username, password);
+
+      PCollection<JdbcDataRecord> output = input.apply(Create.of(query))
```

Comment:
"query" is redundant between Create.of(query) and ReadFn(jdbcOptions which also contains query).
Instead you can do Create.of(jdbcOptions) and make ReadFn argumentless.